lib/dartdoc.dart

```diff
@@ -167,21 +185,170 @@ class DartDoc {
+        continue;
+      }
+      if (fullPath == indexJson) {
+        foundIndex = true;
+        _onCheckProgress.add(fullPath);
+        continue;
+      }
+      if (visited.contains(fullPath)) continue;
+      if (!writtenFiles.contains(fullPath)) {
+        // This isn't a file we wrote (this time); don't claim we did.
+        _warn(package, PackageWarning.unknownFile, fullPath, normalOrigin);
+      } else {
+        _warn(package, PackageWarning.orphanedFile, fullPath, normalOrigin);
+      }
+      _onCheckProgress.add(fullPath);
+    }
+
+    if (!foundIndex) {
+      _warn(package, PackageWarning.brokenLink, indexJson, normalOrigin);
+      _onCheckProgress.add(indexJson);
+    }
+  }
+
+  _doCheck(
+      Package package, String origin, Set<String> visited, String pathToCheck,
+      [String source, String fullPath]) {
+    if (fullPath == null) {
+      fullPath = path.joinAll([origin, pathToCheck]);
+      fullPath = path.normalize(fullPath);
+    }
+
+    File file = new File("$fullPath");
+    if (!file.existsSync()) {
+      _warn(package, PackageWarning.brokenLink, pathToCheck,
+          path.normalize(origin),
+          source: source);
+      _onCheckProgress.add(pathToCheck);
+      return;
+    }
+    Document doc = parse(file.readAsStringSync());
+    Element base = doc.querySelector('base');
+    String baseHref;
+    if (base != null) {
+      baseHref = base.attributes['href'];
+    }
+    List<Element> links = doc.querySelectorAll('a');
+    Iterable<String> stringLinks = links
+        .map((link) => link.attributes['href'])
+        .where((href) => href != null);
+
+    for (String href in stringLinks) {
+      if (!href.startsWith('http') && !href.contains('#')) {
+        var full;
+        if (baseHref != null) {
+          full = '${path.dirname(pathToCheck)}/$baseHref/$href';
+        } else {
+          full = '${path.dirname(pathToCheck)}/$href';
+        }
+        var newPathToCheck = path.normalize(full);
+        String newFullPath = path.joinAll([origin, newPathToCheck]);
+        newFullPath = path.normalize(newFullPath);
+        if (!visited.contains(newFullPath)) {
+          visited.add(newFullPath);
+          _doCheck(package, origin, visited, newPathToCheck, pathToCheck,
+              newFullPath);
+        }
+      }
+    }
+
+    _onCheckProgress.add(pathToCheck);
+    1 + 1;
+    //return;
+  }
+
+  Map<String, Set<ModelElement>> _hrefs;
+
+  /// Don't call this method more than once, and only after you've
+  /// generated all docs for the Package.
+  Future verifyLinks(Package package, String origin) async {
+    assert(_hrefs == null);
+    _hrefs = package.allHrefs;
+
+    Set<String> visited = new Set();
+    String start = 'index.html';
+    visited.add(start);
+    stdout.write('\nvalidating docs');
+    _doCheck(package, origin, visited, start);
+    await _doOrphanCheck(package, origin, visited);
+    // A do nothing line magically makes await _doOrphanCheck actually await?
```

Comment:
It should await anyway; but from comments above, we may be able to remove the async nature here.